pkg/platform/openstack/openstack-heat.go

```diff
@@ -178,6 +175,9 @@ resources:
             availability_zone: {{.AvailabilityZone}}
             {{- end}}
     {{- end}}
+	{{- if .ExistingData }}
```

Comment:
So this rebuilds the volume definition from the new state, but another way to do this would be to include the existing volume definition in the ExistingData object. The choice depends on if there is there a way that 1) the volume definition could change (like the size) or 2) the volumes could be removed/added. If yes, then does that trigger a rebuild of the VM? I do see that it looks like changing the volumes may affect the cloud-init config, which would definitely trigger a VM rebuild.

Looking at clusterinst.proto, it looks like external_volume_size and shared_volume_size are not updatable, so technically volumes should not change. I think then it might be safer to leave the template as it was and just ensure that ExistingData includes any volume definitions. What do you think?

Reply:
I tested and since the attachment VM never gets updated, the volume stays the same and doesn't get blown away(stays with the same ID). I think this is ok, however let me explore keeping it as part of the Existing data to see, which approach is cleaner

Reply:
After looking at how to handle the volumes separately from the VMs it is not straightforward. We should address this in another [issue](https://github.com/edgexr/edge-cloud-platform/issues/335)